sklearn/utils/extmath.py

```diff
@@ -1232,10 +1241,29 @@ def stable_cumsum(arr, axis=None, rtol=1e-05, atol=1e-08):
     out : ndarray
         Array with the cumulative sums along the chosen axis.
     """
-    out = np.cumsum(arr, axis=axis, dtype=np.float64)
-    expected = np.sum(arr, axis=axis, dtype=np.float64)
-    if not np.allclose(
-        out.take(-1, axis=axis), expected, rtol=rtol, atol=atol, equal_nan=True
+    xp, _, device = get_namespace_and_device(arr)
+    if axis is None:
+        # np.cumsum calculates cumsum across flattened array
+        # for axis=None, which is why we flatten here.
+        # (if we don't flatten, cumulative_sum below will raise error)
+        arr = xp.reshape(arr, (-1,))
+        axis = 0
+    max_float_dtype = _max_precision_float_dtype(xp, device)
+    out = xp.cumulative_sum(arr, axis=axis, dtype=max_float_dtype)
+    expected = xp.sum(arr, axis=axis, dtype=max_float_dtype)
+    # workaround to get last element
+    # PyTorch doesn't support this ATM
+    # TODO: remove once issue is fixed
+    # https://github.com/pytorch/pytorch/issues/146211
+    last_elem_idx = xp.asarray(out.shape[axis] - 1)
+    if not xp.all(
+        xpx.isclose(
+            xp.take(out, xp.asarray([last_elem_idx], device=device), axis=axis),
+            expected,
+            rtol=rtol,
+            atol=atol,
+            equal_nan=True,
+        )
     ):
```

Comment:
@lithomas1 seems that this part is not working (CI failure shows no warning emitted for cupy)

Reply:
Am at a lose as to what is causing this failure. The docs do talk about automatic [type promotion](https://docs.cupy.dev/en/stable/user_guide/difference.html#type-promotion):

> CuPy automatically promotes dtypes of [cupy.ndarray](https://docs.cupy.dev/en/v4.5.0/reference/generated/cupy.ndarray.html#cupy.ndarray) s in a function with two or more operands

but the failure occurs even when `dtype` is float64.

Maybe others can shed some light?

Reply:
It still failed in one case. @betatim @lesteve What do you think about this? This warning is not triggered easily.

Reply:
I think I would be in favor of not calling scikit-learn's `stable_cumsum` at all in functions with array API support, and instead directly call `xp.cumsum`.

Once all the code base has been migrated to remove all the calls to `stable_cumsum` we can just deprecate this function. The extra call to `np/xp.sum` is adding some unnecessary overhead for no value for the end user: if the `cumsum` result happens to be unstable, the users cannot do anything about it, so I don't really see the point of this warning.

Reply:
In order to merge this PR should we remove the warning code altogether or should we keep the code but stop testing for it?